File: export_player_template/runtime_choice_availability.js
(function attachCanvasiaRuntimeChoiceAvailability(global) {
  "use strict";

  const CHOICE_CONTINUE_TARGET = "__continue__";
  const CHOICE_SAFETY_OPTION_ID = "__canvasia_choice_safety_continue__";
  const CHOICE_AVAILABILITY_MODES = Object.freeze({
    ALWAYS: "always",
    HIDE_WHEN_FALSE: "hide_when_false",
    DISABLE_WHEN_FALSE: "disable_when_false",
  });
  const CHOICE_AVAILABILITY_MODE_LABELS = Object.freeze({
    always: "始终可选",
    hide_when_false: "条件不满足时隐藏",
    disable_when_false: "条件不满足时锁定",
  });

  function cleanText(value, fallback = "") {
    const text = String(value ?? "").trim();
    return text || fallback;
  }

  function toArray(value) {
    return Array.isArray(value) ? value : [];
  }

  function normalizeChoiceAvailabilityMode(value) {
    const aliases = {
      hide: CHOICE_AVAILABILITY_MODES.HIDE_WHEN_FALSE,
      hidden: CHOICE_AVAILABILITY_MODES.HIDE_WHEN_FALSE,
      disable: CHOICE_AVAILABILITY_MODES.DISABLE_WHEN_FALSE,
      disabled: CHOICE_AVAILABILITY_MODES.DISABLE_WHEN_FALSE,
      locked: CHOICE_AVAILABILITY_MODES.DISABLE_WHEN_FALSE,
    };
    const candidate = aliases[cleanText(value)] ?? cleanText(value, CHOICE_AVAILABILITY_MODES.ALWAYS);
    return Object.hasOwn(CHOICE_AVAILABILITY_MODE_LABELS, candidate)
      ? candidate
      : CHOICE_AVAILABILITY_MODES.ALWAYS;
  }

  function getChoiceAvailabilityRules(option = {}) {
    return toArray(option.choiceAvailabilityWhen ?? option.availabilityWhen ?? option.when);
  }

  function getChoiceLockedReason(option = {}, fallback = "条件尚未满足") {
    return cleanText(option.choiceLockedReason ?? option.lockedReason, fallback);
  }

  function evaluateChoiceAvailabilityRule(rule = {}, variableValues = {}, options = {}) {
    if (typeof options.evaluateRule === "function") {
      return Boolean(options.evaluateRule(rule, variableValues, options));
    }
    const conditionTools = options.conditionTools ?? global.CanvasiaRuntimeConditions;
    if (typeof conditionTools?.evaluateConditionRule !== "function") {
      return false;
    }
    return Boolean(conditionTools.evaluateConditionRule(rule, variableValues, options.conditionOptions ?? options));
  }

  function inspectChoiceOptionAvailability(option = {}, variableValues = {}, options = {}) {
    const mode = normalizeChoiceAvailabilityMode(option.choiceAvailabilityMode ?? option.availabilityMode);
    const rules = getChoiceAvailabilityRules(option);
    const ruleResults = rules.map((rule) => evaluateChoiceAvailabilityRule(rule, variableValues, options));
    const matched = mode === CHOICE_AVAILABILITY_MODES.ALWAYS
      ? true
      : rules.length > 0 && ruleResults.every(Boolean);
    const visible = mode !== CHOICE_AVAILABILITY_MODES.HIDE_WHEN_FALSE || matched;
    const enabled = visible && (mode !== CHOICE_AVAILABILITY_MODES.DISABLE_WHEN_FALSE || matched);

    return {
      mode,
      modeLabel: CHOICE_AVAILABILITY_MODE_LABELS[mode],
      rules,
      ruleResults,
      matched,
      visible,
      enabled,
      lockedReason: enabled ? "" : getChoiceLockedReason(option),
    };
  }

  function createChoiceSafetyOption(options = {}) {
    return {
      id: CHOICE_SAFETY_OPTION_ID,
      text: cleanText(options.safetyText, "继续剧情（安全兜底）"),
      gotoSceneId: CHOICE_CONTINUE_TARGET,
      effects: [],
      choiceAvailabilityMode: CHOICE_AVAILABILITY_MODES.ALWAYS,
      choiceAvailabilityMatched: true,
      choiceVisible: true,
      choiceEnabled: true,
      choiceLockedReason: "",
      isChoiceSafetyFallback: true,
    };
  }

  function resolveChoiceOptions(choiceOptions = [], variableValues = {}, options = {}) {
    const allOptions = toArray(choiceOptions).map((option, index) => {
      const availability = inspectChoiceOptionAvailability(option, variableValues, options);
      return {
        ...option,
        id: cleanText(option?.id, `choice_option_${index + 1}`),
        choiceAvailabilityMode: availability.mode,
        choiceAvailabilityMatched: availability.matched,
        choiceAvailabilityRuleResults: availability.ruleResults,
        choiceVisible: availability.visible,
        choiceEnabled: availability.enabled,
        choiceLockedReason: availability.lockedReason,
        isChoiceSafetyFallback: false,
      };
    });
    const authoredVisibleOptions = allOptions.filter((option) => option.choiceVisible);
    const selectableOptions = authoredVisibleOptions.filter((option) => option.choiceEnabled);
    const allUnavailable = selectableOptions.length === 0;
    const safetyOption = allUnavailable && options.includeSafetyFallback !== false
      ? createChoiceSafetyOption(options)
      : null;
    const runtimeOptions = safetyOption ? [...authoredVisibleOptions, safetyOption] : authoredVisibleOptions;

    return {
      allOptions,
      authoredVisibleOptions,
      selectableOptions,
      runtimeOptions,
      safetyOption,
      allUnavailable,
      hiddenCount: allOptions.filter((option) => !option.choiceVisible).length,
      lockedCount: authoredVisibleOptions.filter((option) => !option.choiceEnabled).length,
    };
  }

  function isChoiceOptionSelectable(option = {}) {
    return option.choiceVisible !== false && option.choiceEnabled !== false;
  }

  function findSelectableChoiceIndex(options = [], startIndex = 0, direction = 1) {
    const items = toArray(options);
    if (!items.length) {
      return -1;
    }
    const step = Number(direction) < 0 ? -1 : 1;
    const safeStart = Number.isFinite(Number(startIndex)) ? Number(startIndex) : 0;
    for (let offset = 0; offset < items.length; offset += 1) {
      const index = ((safeStart + offset * step) % items.length + items.length) % items.length;
      if (isChoiceOptionSelectable(items[index])) {
        return index;
      }
    }
    return -1;
  }

  global.CanvasiaRuntimeChoiceAvailability = Object.freeze({
    CHOICE_CONTINUE_TARGET,
    CHOICE_SAFETY_OPTION_ID,
    CHOICE_AVAILABILITY_MODES,
    CHOICE_AVAILABILITY_MODE_LABELS,
    normalizeChoiceAvailabilityMode,
    getChoiceAvailabilityRules,
    getChoiceLockedReason,
    evaluateChoiceAvailabilityRule,
    inspectChoiceOptionAvailability,
    createChoiceSafetyOption,
    resolveChoiceOptions,
    isChoiceOptionSelectable,
    findSelectableChoiceIndex,
  });
})(typeof window !== "undefined" ? window : globalThis);

File: native_runtime/runtime_choice_availability.py
from __future__ import annotations

from collections.abc import Callable, Iterable
from typing import Any


CHOICE_CONTINUE_TARGET = "__continue__"
CHOICE_SAFETY_OPTION_ID = "__canvasia_choice_safety_continue__"
CHOICE_AVAILABILITY_ALWAYS = "always"
CHOICE_AVAILABILITY_HIDE = "hide_when_false"
CHOICE_AVAILABILITY_DISABLE = "disable_when_false"
CHOICE_AVAILABILITY_MODES = {
    CHOICE_AVAILABILITY_ALWAYS,
    CHOICE_AVAILABILITY_HIDE,
    CHOICE_AVAILABILITY_DISABLE,
}


def _clean_text(value: Any, fallback: str = "") -> str:
    text = str(value or "").strip()
    return text or fallback


def normalize_choice_availability_mode(value: Any) -> str:
    aliases = {
        "hide": CHOICE_AVAILABILITY_HIDE,
        "hidden": CHOICE_AVAILABILITY_HIDE,
        "disable": CHOICE_AVAILABILITY_DISABLE,
        "disabled": CHOICE_AVAILABILITY_DISABLE,
        "locked": CHOICE_AVAILABILITY_DISABLE,
    }
    candidate = aliases.get(_clean_text(value), _clean_text(value, CHOICE_AVAILABILITY_ALWAYS))
    return candidate if candidate in CHOICE_AVAILABILITY_MODES else CHOICE_AVAILABILITY_ALWAYS


def get_choice_availability_rules(option: dict[str, Any] | None) -> list[dict[str, Any]]:
    source = option or {}
    rules = source.get("choiceAvailabilityWhen", source.get("availabilityWhen", source.get("when", [])))
    return [rule for rule in rules if isinstance(rule, dict)] if isinstance(rules, list) else []


def inspect_choice_option_availability(
    option: dict[str, Any] | None,
    evaluate_rules: Callable[[list[dict[str, Any]]], bool] | None = None,
) -> dict[str, Any]:
    source = option or {}
    mode = normalize_choice_availability_mode(source.get("choiceAvailabilityMode", source.get("availabilityMode")))
    rules = get_choice_availability_rules(source)
    matched = True if mode == CHOICE_AVAILABILITY_ALWAYS else bool(rules and evaluate_rules and evaluate_rules(rules))
    visible = mode != CHOICE_AVAILABILITY_HIDE or matched
    enabled = visible and (mode != CHOICE_AVAILABILITY_DISABLE or matched)
    return {
        "mode": mode,
        "rules": rules,
        "matched": matched,
        "visible": visible,
        "enabled": enabled,
        "lockedReason": "" if enabled else _clean_text(
            source.get("choiceLockedReason", source.get("lockedReason")),
            "条件尚未满足",
        ),
    }


def create_choice_safety_option(text: str = "继续剧情（安全兜底）") -> dict[str, Any]:
    return {
        "id": CHOICE_SAFETY_OPTION_ID,
        "text": _clean_text(text, "继续剧情（安全兜底）"),
        "gotoSceneId": CHOICE_CONTINUE_TARGET,
        "effects": [],
        "choiceAvailabilityMode": CHOICE_AVAILABILITY_ALWAYS,
        "choiceAvailabilityMatched": True,
        "choiceVisible": True,
        "choiceEnabled": True,
        "choiceLockedReason": "",
        "isChoiceSafetyFallback": True,
    }


def resolve_runtime_choice_options(
    choice_options: Iterable[dict[str, Any]] | None,
    evaluate_rules: Callable[[list[dict[str, Any]]], bool] | None = None,
    *,
    include_safety_fallback: bool = True,
    safety_text: str = "继续剧情（安全兜底）",
) -> dict[str, Any]:
    all_options: list[dict[str, Any]] = []
    for index, raw_option in enumerate(choice_options or []):
        option = dict(raw_option) if isinstance(raw_option, dict) else {}
        availability = inspect_choice_option_availability(option, evaluate_rules)
        option.update(
            {
                "id": _clean_text(option.get("id"), f"choice_option_{index + 1}"),
                "choiceAvailabilityMode": availability["mode"],
                "choiceAvailabilityMatched": availability["matched"],
                "choiceVisible": availability["visible"],
                "choiceEnabled": availability["enabled"],
                "choiceLockedReason": availability["lockedReason"],
                "isChoiceSafetyFallback": False,
            }
        )
        all_options.append(option)

    authored_visible_options = [option for option in all_options if option["choiceVisible"]]
    selectable_options = [option for option in authored_visible_options if option["choiceEnabled"]]
    all_unavailable = not selectable_options
    safety_option = create_choice_safety_option(safety_text) if all_unavailable and include_safety_fallback else None
    runtime_options = [*authored_visible_options, safety_option] if safety_option else authored_visible_options
    return {
        "allOptions": all_options,
        "authoredVisibleOptions": authored_visible_options,
        "selectableOptions": selectable_options,
        "runtimeOptions": runtime_options,
        "safetyOption": safety_option,
        "allUnavailable": all_unavailable,
        "hiddenCount": sum(not option["choiceVisible"] for option in all_options),
        "lockedCount": sum(not option["choiceEnabled"] for option in authored_visible_options),
    }


def is_choice_option_selectable(option: dict[str, Any] | None) -> bool:
    source = option or {}
    return source.get("choiceVisible") is not False and source.get("choiceEnabled") is not False


def find_selectable_choice_index(options: list[dict[str, Any]] | None, start_index: int = 0, direction: int = 1) -> int:
    items = options or []
    if not items:
        return -1
    step = -1 if direction < 0 else 1
    for offset in range(len(items)):
        index = (start_index + offset * step) % len(items)
        if is_choice_option_selectable(items[index]):
            return index
    return -1
